src/Compiler/NodeVisitor/AddFormClassesNodeVisitor.php

```diff
@@ -0,0 +1,181 @@
+                return null;
+            }
+            $dimFetch = $assign->expr;
+
+            if (!$dimFetch->dim instanceof String_) {
+                return null;
+            }
+            $formNameString = $dimFetch->dim;
+            $formName = $formNameString->value;
+
+            $formClassName = $this->formClassNames[$formName] ?? null;
+            if ($formClassName === null) {
+                return null;
+            }
+            $this->actualForm = $this->forms[$formName] ?? null;
+            return new Assign(new Variable('form'), new New_(new Name($formClassName)));
+        } elseif ($node instanceof StaticCall) {
+            if ($this->nameResolver->resolve($node) === 'renderFormEnd') {
+                $node->args[0] = new Arg(new Variable('form'));
+                $this->actualForm = null;
+                return $node;
+            }
+        } elseif ($node instanceof ArrayDimFetch) {
+            if ($this->actualForm === null) {
+                return null;
+            }
+            if (!$node->dim instanceof String_) {
+                return null;
+            }
+
+            if (!$node->var instanceof FuncCall) {
+                return null;
+            }
+
+            if ($this->nameResolver->resolve($node->var) !== 'end') {
+                return null;
+            }
+
+            $fieldName = $node->dim->value;
+            $formField = $this->actualForm->getFormField($fieldName);
+            if ($formField === null) {
+                $rootParentNode = $node;
+                while (true) {
+                    $parentNode = $rootParentNode->getAttribute('parent');
+                    if ($parentNode === null) {
+                        break;
+                    }
+
+                    $rootParentNode = $parentNode;
+
+                    if ($rootParentNode instanceof Echo_ || $rootParentNode instanceof If_) {
+                        break;
+                    }
+                }
+                $this->errorFieldNodes[] = [
+                    'node' => $rootParentNode,
+                    'field' => $fieldName,
+                ];
+                return null;
+            }
+
+            $node->var = new Variable('form');
+            return $node;
+        }
+
+        return null;
+    }
+
+    public function leaveNode(Node $node)
+    {
+        foreach ($this->errorFieldNodes as $errorFieldNode) {
+            if ($errorFieldNode['node'] === $node) {
+                $error = new Error('Form field with name "' . $errorFieldNode['field'] . '" probably does not exist.');
+                $errorNode = $error->toNode();
+                $errorNode->setAttributes($node->getAttributes());
+                return $errorNode;
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * @param Node[] $nodes
+     * @return Node[]
+     */
+    public function afterTraverse(array $nodes): array
+    {
+        $componentType = '\Nette\ComponentModel\IComponent';
```

Comment:
Would not it be better to expect Form control here instead of generic component? It almost always would be control.

Reply:
This is just compatibility with nette itself, unfortunatelly you can add any Nette\ComponentModel\IComponent to Form / Container. But I would prefere some Form\BaseControl here, too.

Reply:
I think we can reasonably expect that when someone uses `{input}` or `{label}` he is trying to access `BaseControl` and in other cases this definition is not even used I think.

Reply:
that would be great, I'll change it to BaseControl than...